File: php/core/bt_api_basic.php
<?php
/* 
 * btTelco Basic REST API - PHP Implementation
 * 
 * Version 1.21
 *
 * btTelco
 * A bitTrust Service.
 * Identikrecord Group
 * Copyright 2013 (C) All rights reserved.
 * 
 */


class BtRestAPI {
	private $_api_url;
	private $_acct_name;
	private $_acct_key;

	private function _request_url($url) {
		$ch = curl_init();

		curl_setopt($ch, CURLOPT_URL, $url);
		curl_setopt($ch, CURLOPT_FOLLOWLOCATION, true);
		curl_setopt($ch, CURLOPT_RETURNTRANSFER, true);
		curl_setopt($ch, CURLOPT_SSL_VERIFYHOST, false);
		curl_setopt($ch, CURLOPT_SSL_VERIFYPEER, false);

		if (curl_exec($ch) === false) {
			error_log('btTelco PHP API: curl_exec(): cURL error: ' . curl_error($ch));
			return false;
		}

		$res = curl_multi_getcontent($ch);

		curl_close($ch);

		if (($res_json = json_decode($res, true)) == NULL) {
			error_log('btTelco PHP API: json_decode(): Unable to decode JSON response');
			return false;
		}

		return $res_json;
	}

	public function __construct($api_url, $acct_name, $acct_key) {
		$this->_api_url = $api_url . '/index.php';
		$this->_acct_name = $acct_name;
		$this->_acct_key = $acct_key;
	}

	public function request_sms($op, $to, $msg, $from = false, $timestamp = false, $mid = false) {
		$url = $this->_api_url . '/sms/';

		if ($from === false)
			$from = 'false';

		if ($timestamp === false)
			$timestamp = 'false';

		if ($mid === false) {
			$url .= $op . '/' .
				$this->_acct_name . '/' .
				$this->_acct_key . '/' .
				$to . '/' .
				rawurlencode($msg) . '/' .
				$from . '/' .
				$timestamp;
		} else {
			$url .= $op . '/' .
				$this->_acct_name . '/' .
				$this->_acct_key . '/' .
				$mid;
		}

		return $this->_request_url($url);
	}

	public function request_ussd($op, $to, $msg, $from = false, $timestamp = false, $mid = false) {
		$url = $this->_api_url . '/ussd/';

		if ($from === false)
			$from = 'false';

		if ($timestamp === false)
			$timestamp = 'false';

		if ($mid === false) {
			$url .= $op . '/' .
				$this->_acct_name . '/' .
				$this->_acct_key . '/' .
				$to . '/' .
				rawurlencode($msg) . '/' .
				$from . '/' .
				$timestamp;
		} else {
			$url .= $op . '/' .
				$this->_acct_name . '/' .
				$this->_acct_key . '/' .
				$mid;
		}

		return $this->_request_url($url);
	}

	public function request_tts($op, $to, $msg, $from = false, $language = false, $gender = false, $timestamp = false, $mid = false) {
		$url = $this->_api_url . '/tts/';

		if ($from === false)
			$from = 'false';

		if ($language === false)
			$language = 'false';

		if ($gender === false)
			$gender = 'false';

		if ($timestamp === false)
			$timestamp = 'false';

		if ($mid === false) {
			$url .= $op . '/' .
				$this->_acct_name . '/' .
				$this->_acct_key . '/' .
				$to . '/' .
				rawurlencode($msg) . '/' .
				$from . '/' .
				$language . '/' .
				$gender . '/' .
				$timestamp;
		} else {
			$url .= $op . '/' .
				$this->_acct_name . '/' .
				$this->_acct_key . '/' .
				$mid;
		}

		return $this->_request_url($url);
	}
}

class BtSmsAPI {
	private $_rest;

	public function __construct($api_url, $acct_name, $acct_key) {
		$this->_rest = new BtRestAPI($api_url, $acct_name, $acct_key);
	}

	public function send($to, $msg, $from = false, $timestamp = false) {
		return $this->_rest->request_sms('send', $to, $msg, $from, $timestamp);
	}

	public function report($mid) {
		return $this->_rest->request_sms('report', false, false, false, false, $mid);
	}
}

class BtUssdAPI {
	private $_rest;

	public function __construct($api_url, $acct_name, $acct_key) {
		$this->_rest = new BtRestAPI($api_url, $acct_name, $acct_key);
	}

	public function send($to, $msg, $from = false, $timestamp = false) {
		return $this->_rest->request_ussd('send', $to, $msg, $from, $timestamp);
	}

	public function report($mid) {
		return $this->_rest->request_ussd('report', false, false, false, false, $mid);
	}
}

class BtTtsAPI {
	private $_rest;

	public function __construct($api_url, $acct_name, $acct_key) {
		$this->_rest = new BtRestAPI($api_url, $acct_name, $acct_key);
	}

	public function send($to, $msg, $from = false, $language = false, $gender = false, $timestamp = false) {
		return $this->_rest->request_tts('send', $to, $msg, $from, $language, $gender, $timestamp);
	}

	public function report($mid) {
		return $this->_rest->request_tts('report', false, false, false, false, false, false, $mid);
	}
}

class BtAPIBasic {
	public $sms;
	public $ussd;
	public $tts;

	public function __construct($api_url, $acct_name, $acct_key) {
		$this->acct_name = $acct_name;
		$this->acct_key = $acct_key;
		$this->sms = new BtSmsAPI($api_url, $acct_name, $acct_key);
		$this->sms = new BtUssdAPI($api_url, $acct_name, $acct_key);
		$this->tts = new BtTtsAPI($api_url, $acct_name, $acct_key);
	}
}

File: python/core/bt_api_basic.py
#!/usr/bin/python -P

# 
# btTelco Basic REST API - Python Implementation
# 
# Version 1.21
#
# btTelco - http://www.bttelco.com
#
# A bitTrust Service.
# http://www.bittrust.net
# Identikrecord Group
# Copyright 2013 (C) All rights reserved.
# 


import urllib
import json
import syslog

class BtRestAPI:
	def __init__(self, api_url, acct_name, acct_key):
		self.api_url = api_url + '/index.php'
		self.acct_name = acct_name
		self.acct_key = acct_key

	def _request_url(self, url):
		print url
		try:
			r = urllib.urlopen(url)
		except:
			syslog.syslog('BtRestAPI._request_url(): urllib2.urlopen(\"%s\"): Failed' % url)
			return False

		try:
			res_json = json.load(r)
		except:
			syslog.syslog('BtRestAPI._request_url(): json.load(): Failed')
			return False

		return res_json

	def request_sms(self, op, to, msg, sender = False, timestamp = False, mid = False):
		url = self.api_url + '/sms'

		if sender == False:
			sender = 'false'

		if timestamp == False:
			timestamp = 'false'

		if mid == False:
			url = "%s/%s/%s/%s/%s/%s/%s/%s" % (url, op, self.acct_name, self.acct_key, to, urllib.quote(msg), sender, timestamp)
		else:
			url = "%s/%s/%s/%s/%s" % (url, op, self.acct_name, self.acct_key, mid)

		return self._request_url(url)

	def request_ussd(self, op, to, msg, sender = False, timestamp = False, mid = False):
		url = self.api_url + '/ussd'

		if sender == False:
			sender = 'false'

		if timestamp == False:
			timestamp = 'false'

		if mid == False:
			url = "%s/%s/%s/%s/%s/%s/%s/%s" % (url, op, self.acct_name, self.acct_key, to, urllib.quote(msg), sender, timestamp)
		else:
			url = "%s/%s/%s/%s/%s" % (url, op, self.acct_name, self.acct_key, mid)

		return self._request_url(url)

	def request_tts(self, op, to, msg, from_num = False, language = False, gender = False, timestamp = False, mid = False):
		url = self.api_url + '/tts'

		if from_num == False:
			sender = 'false'

		if language == False:
			language = 'false'

		if gender == False:
			gender = 'false'

		if timestamp == False:
			timestamp = 'false'

		if mid == False:
			url = "%s/%s/%s/%s/%s/%s/%s/%s/%s/%s" % (url, op, self.acct_name, self.acct_key, to, urllib.quote(msg), sender, language, gender, timestamp)
		else:
			url = "%s/%s/%s/%s/%s" % (url, op, self.acct_name, self.acct_key, mid)

		return self._request_url(url)

class BtSmsAPI:
	def __init__(self, api_url, acct_name, acct_key):
		self._rest = BtRestAPI(api_url, acct_name, acct_key)

	def send(self, to, msg, sender = False, timestamp = False):
		return self._rest.request_sms('send', to, msg, sender, timestamp)

	def report(self, mid):
		return self._rest.request_sms('report', False, False, False, False, mid)

class BtUssdAPI:
	def __init__(self, api_url, acct_name, acct_key):
		self._rest = BtRestAPI(api_url, acct_name, acct_key)

	def send(self, to, msg, sender = False, timestamp = False):
		return self._rest.request_ussd('send', to, msg, sender, timestamp)

	def report(self, mid):
		return self._rest.request_ussd('report', False, False, False, False, mid)

class BtTtsAPI:
	def __init__(self, api_url, acct_name, acct_key):
		self._rest = BtRestAPI(api_url, acct_name, acct_key)

	def send(self, to, msg, from_num = False, language = False, gender = False, timestamp = False):
		return self._rest.request_tts('send', to, msg, from_num, language, gender, timestamp)

	def report(self, mid):
		return self._rest.request_tts('report', False, False, False, False, False, False, mid)

class BtAPIBasic:
	def __init__(self, api_url, acct_name, acct_key):
		self.acct_name = acct_name
		self.acct_key = acct_key
		self.sms = BtSmsAPI(api_url, acct_name, acct_key)
		self.ussd = BtUssdAPI(api_url, acct_name, acct_key)
		self.tts = BtTtsAPI(api_url, acct_name, acct_key)

